release/ray_release/command_runner/anyscale_job_runner.py

```diff
@@ -0,0 +1,265 @@
+            logger.info(f"Output: {output_json}")
+            workload_status_code = output_json["return_code"]
+            workload_time_taken = output_json["workload_time_taken"]
+            prepare_return_codes = output_json["prepare_return_codes"]
+            last_prepare_time_taken = output_json["last_prepare_time_taken"]
+
+            if prepare_return_codes and prepare_return_codes[-1] != 0:
+                if prepare_return_codes[-1] == TIMEOUT_RETURN_CODE:
+                    raise PrepareCommandTimeout(
+                        "Prepare command timed out after "
+                        f"{last_prepare_time_taken} seconds."
+                    )
+                raise PrepareCommandError(
+                    f"Prepare command '{self.prepare_commands[-1]}' returned "
+                    f"non-success status: {prepare_return_codes[-1]} with error:"
+                    f"\n{error}\n"
+                )
+        else:
+            raise JobNoLogsError("Could not obtain logs for the job.")
+
+        if workload_status_code == TIMEOUT_RETURN_CODE:
+            if is_long_running:
+                # Expected - treat as success.
+                return
+
+            raise TestCommandTimeout(
+                f"Command timed out after {workload_time_taken} seconds."
+            )
+
+        if workload_status_code is not None and workload_status_code != 0:
+            raise TestCommandError(
+                f"Command returned non-success status: {workload_status_code} with "
+                f"error:\n{error}\n"
+            )
+
+        if job_status_code == -1:
+            raise JobStartupFailed(
+                "Job returned non-success state: 'OUT_OF_RETRIES' "
+                "(command has not been ran or no logs could have been obtained) "
+                f"with error:\n{error}\n"
+            )
+
+    @property
+    def command_env(self):
+        env = super().command_env
+        # Make sure we don't buffer stdout so we don't lose any logs.
+        env["PYTHONUNBUFFERED"] = "1"
+        return env
+
+    def run_command(
+        self,
+        command: str,
+        env: Optional[Dict] = None,
+        timeout: float = 3600.0,
+        is_long_running: bool = False,
+    ) -> float:
+        prepare_command_strs = []
+        prepare_command_timeouts = []
+        for prepare_command, prepare_env, prepare_timeout in self.prepare_commands:
+            prepare_env = self.get_full_command_env(prepare_env)
+
+            if prepare_env:
+                env_str = " ".join(f"{k}={v}" for k, v in prepare_env.items()) + " "
+            else:
+                env_str = ""
+            prepare_command_strs.append(f"{env_str} {prepare_command}")
+            prepare_command_timeouts.append(prepare_timeout)
+
+        prepare_commands = " ".join(shlex.quote(str(x)) for x in prepare_command_strs)
+        prepare_commands_timeouts = " ".join(
+            shlex.quote(str(x)) for x in prepare_command_timeouts
+        )
```

Comment:
I get what is happening here but can we add some in-line comments? Including an example of how a prepare command looks like and how it will be rendered in the command